### PR TITLE
Add Nuxt, ESM, Typescript

### LINKED ISSUE #1232: Prototype Pollution Check
https://github.com/wg-easy/wg-easy/blob/1416613cc671c0bb86eca6708740482fa965806a/src/lib/Server.js#L205-L208

Is there really a issue with prototype polution? 
The key of an object has to be `__proto__` but this creates a object with `{ clientId: '__proto__' }`

Maybe the better solution is to check if the clientId is a valid uuid with something like zod

### PR DESCRIPTION
**Pull request by kaaax0815**
Supersedes / closes #937 

Still very much in development

- [x] move all api functions
- [x] improve ofetch errors
- [x] validate using zod (closes #1232)
- [x] split into components
- - [x] fix apexchart 
- - [x] fix all linting errors
- - [x] fix all typescript errors
- - [x] use or create icon library
- [x] fix styling issues
- [ ] update all deployment methods
- [ ] test
- [x] improve tooling (lint, format, typecheck)
- [x] re-implement features/changes implemented in #937
- [x] ~~squash some commits before merge for cleaner history~~ (whole PR has to be squash merged)

Tracker (stuff that should be addressed at some point):
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2274071614
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2275309933
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2275862683
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2277204774
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2284003003
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2286220939
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2286319680
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2296376876
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2298959590
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2301232639
- https://github.com/wg-easy/wg-easy/pull/1244#issuecomment-2306623943

### CODE REVIEW

**Comment by kaaax0815** [comment]
As of now. Everything works with no apparent regression. Only the session guard doesn't work just yet

**Comment by tetuaoro** [comment]
Hi,

need this pr before merging my pr

**Comment by pheiduck** [comment]
Well, the Linter is broken due this changes here. Should be fixed to before merge, as well our package update bot.

**Comment by kaaax0815** [comment]
> Any reason to move to `pnpm`?

Can revert, but its insanely fast

**Comment by pheiduck** [comment]
> Can revert, but its insanely fast

I prefer a revert do it looks like time wasting.

**Comment by kaaax0815** [comment]
> I prefer a revert do it looks like time wasting.

Works now. If you prefer I can still revert

**Comment by pheiduck** [comment]
> Works now. If you prefer I can still revert

Looks fine now.

**Comment by kaaax0815** [comment]
Works fine. Would appreciate help with splitting into components

**Comment by tetuaoro** [comment]
> Any reason to move to `pnpm`?

@kaaax0815 I use also `pnpm`, but I prefer to revert to `npm` because some devs use `yarn`. At least, we all agree to use `npm`.

**Comment by tetuaoro** [comment]
> Works fine. Would appreciate help with splitting into components

I'll help

**Comment by kaaax0815** [comment]
Instead of polling the client list (and stats) like we do now. We could also do it using Nitros Websockets. What's your opinion?

**Comment by tetuaoro** [comment]
I did already and waiting for feedback. I did not push against this pr.

**Comment by tetuaoro** [comment]
> (which doesnt work as you already have a fork)

Ah, if I add a new remote, do you think I can do it ?

**Comment by tetuaoro** [comment]
> My bad. I didn't know you could make a pr from one fork to another fork. Please see the comment on [kaaax0815#1](https://github.com/kaaax0815/wg-easy/pull/1)

yes just add a new remote `git remote add ...`.

**Comment by tetuaoro** [comment]
using nuxt will solve the cache problem by files naming #1263

**Comment by tetuaoro** [comment]
> @tetuaoro Are you working on splitting it more?

yes but I don't have permission to push to your repo

**Comment by kaaax0815** [comment]
Make a PR against feat/nuxt

**Comment by pheiduck** [comment]
Thanks you all for this nice work, due I want to merge this first before everything else.
Do we should use a Beta Branch first?

I know it's a draft, just to let you guys know whats the next path ;)

**Comment by kaaax0815** [comment]
I think a little more splitting would be optimal and completing the tasks.
after that all the planned features in https://github.com/wg-easy/wg-easy/discussions/1179 could be added in seperate pr.

because this then isn't really a draft, as everything works. but splitting would result in big conflicts so doing it first makes most sense imo

File: src/layouts/Footer.vue
Comment (cany748) [inline src/layouts/Footer.vue:1]: This is a misuse of Layouts. If you want to separate Header and Footer into separate components, use regular components, not Layouts.

For example:
components/Layout/LayoutHeader.vue
components/Layout/LayoutFooter.vue

File: src/server/utils/WireGuard.ts [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+  async backupConfiguration() {
+    debug('Starting configuration backup.');
+    const config = await this.getConfig();
+    const backup = JSON.stringify(config, null, 2);
+    debug('Configuration backup completed.');
+    return backup;
+  }
+
+  // Shutdown wireguard
+  async Shutdown() {
```
Comment (cany748) [inline src/server/utils/WireGuard.ts:389]: this method is not called anywhere

**Comment by tetuaoro** [comment]
mr-4 ??

**Comment by tetuaoro** [comment]
> > mr-4 ??
> 
> There is no reference of `mr-4` in `master` or `feat/nuxt`

add  `space-x-*` or `gap-y-*`

**Comment by tetuaoro** [comment]
> the header is now also shown on the login page. any design ideas?

move `layouts/header` to `components` or remove `<h1>...</h1>` from `login.vue` ?!!

**Comment by tetuaoro** [comment]
> > add `space-x-*` or `gap-y-*`
> 
> I know how to fix it. But I wanted to stay as close as possible to the original because of that I asked how it was done on the old one

ok

File: src/layouts/Footer.vue
Comment (tetuaoro) [reply src/layouts/Footer.vue:1]: > This is a misuse of Layouts. If you want to separate Header and Footer into separate components, use regular components, not Layouts.
> 
> For example: components/Layout/LayoutHeader.vue components/Layout/LayoutFooter.vue

i didn't see that, ok

**Comment by tetuaoro** [comment]
fixed [pr](https://github.com/kaaax0815/wg-easy/pull/3)

**Comment by pheiduck** [comment]
> so the best solution is the remove the header from the login page. should the dark mode switch be available on the login page?

Yes, that's ok the user can then already decide between the 3 theme modes.

> and one to be able to show the password would be nice

Hm, for my case it would useless due I have stored my Password in a Password Manager.
And if you know that Windows 11 24H2 can read everything from your Screen and may upload this to microsoft cloud. Show Password is bad.

**Comment by pheiduck** [comment]
> And if you think you mistyped and have to start over again its a bit of a pain

good point.

**Comment by pheiduck** [comment]
Will take a look at the Weekend or later...
Well, discoverd now @kaaax0815 and @offizium-berndstorath 1 Person lul xD

File: src/components/ui/Modal.vue
Comment (tetuaoro) [inline src/components/ui/Modal.vue:1]: remove empty file?! and split more after merge to safe branch, idk

File: src/layouts/Footer.vue [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        class="hover:underline"
+        href="https://github.com/sponsors/WeeJeWel"
+        target="_blank"
+        >{{ $t('donate') }}</a
+      >
+    </p>
+  </footer>
+</template>
+
+<script setup lang="ts"></script>
```
Comment (tetuaoro) [inline src/layouts/Footer.vue:35]: empty

**Review by tetuaoro — COMMENTED** [review]
great, now we can add more features easily and with a smile @thx @kaaax0815

File: src/components/ui/Modal.vue
Comment (kaaax0815) [reply src/components/ui/Modal.vue:1]: Yes. I think splitting into most basic components is a lot of work for not that much benefit. If you do your ui change with settings etc and notice reuse components just split it then

File: src/server/utils/WireGuard.ts (hunk 1 above)
Comment (cany748) [reply src/server/utils/WireGuard.ts:391]: What's fixed here? It's not a proper use of the Layouts folder.
See: https://github.com/nuxt-ui-pro/saas/tree/main/app/layouts

**Comment by pheiduck** [comment]
Maybe it will not merged due @WeeJeWel is asking what’s the benefit for the users with this change.

**Comment by tetuaoro** [comment]
> Maybe it will not merged due @WeeJeWel is asking what’s the benefit for the users with this change?

Haha not for users, for us

File: src/layouts/Footer.vue
Comment (cany748) [reply src/layouts/Footer.vue:1]: What's fixed here? It's not a proper use of the Layouts folder.
See: https://github.com/nuxt-ui-pro/saas/tree/main/app/layouts

**Comment by pheiduck** [comment]
> Haha not for users, for us

He has heard this so many times in his career as a developer that a new codebase/framework helps developers add new features more easily.

**Comment by cany748** [comment]
> Maybe it will not merged due @WeeJeWel is asking what’s the benefit for the users with this change?

If done right, users will benefit from:
1) the small size of the container
2) for those who don't use docker they won't need to install npm dependencies to run it
3) vue 3 is smaller and faster than vue 2

**Comment by kaaax0815** [comment]
Benefits.
Typescript if you change some return type of the API you instantly get errors in your frontend and know where and what to change.
Esm cleaner syntax, potential better tree shaking
Due to better Docker file faster build time and hmr
composition API and script setup for way cleaner syntax
Components for easy and performant reusability

File: src/layouts/Footer.vue
Comment (kaaax0815) [reply src/layouts/Footer.vue:1]: As stated here https://nuxt.com/docs/guide/directory-structure/layouts it's for common ui patterns. Headers and footers are by definition common ui patterns as they are shown on (almost) all pages

**Comment by pheiduck** [comment]
I will wait for @WeeJeWel opinion on this and then represent it accordingly.

**Comment by kaaax0815** [comment]
I'll wait for #1299, merge with master and accept this pr

**Comment by kaaax0815** [comment]
we should do something about the bundle size.
vue3-apexcharts uses 510kB which is 56.03% of the bundle

**Comment by tetuaoro** [comment]
> we should do something about the bundle size. vue3-apexcharts uses 510kB which is 56.03% of the bundle

Wow really ?! Haha

**Comment by kaaax0815** [comment]
now even if you have no javascript you can see the client list

**Comment by pheiduck** [comment]
Published a Dev build: `ghcr.io/wg-easy/wg-easy:development`

**Comment by kaaax0815** [comment]
There are still errors that I need to address. But the overall system still works. Hoping to help @tetuaoro with the new ui design and onboarding

**Pull request merged**